clang/lib/Sema/SemaDecl.cpp

```diff
@@ -18717,16 +18717,14 @@ Sema::FunctionEmissionStatus Sema::getEmissionStatus(FunctionDecl *FD,
                                                      bool Final) {
   assert(FD && "Expected non-null FunctionDecl");
 
-  // SYCL functions can be template, so we check if they have appropriate
-  // attribute prior to checking if it is a template.
-  if (LangOpts.SYCLIsDevice &&
-      (FD->hasAttr<SYCLDeviceAttr>() || FD->hasAttr<SYCLKernelAttr>()))
-    return FunctionEmissionStatus::Emitted;
-
   // Templates are emitted when they're instantiated.
   if (FD->isDependentContext())
     return FunctionEmissionStatus::TemplateDiscarded;
 
+  if (LangOpts.SYCLIsDevice &&
+      (FD->hasAttr<SYCLDeviceAttr>() || FD->hasAttr<SYCLKernelAttr>()))
+    return FunctionEmissionStatus::Emitted;
+
```

Comment:
Could you please explain what is the reason to move the codes here from Line#18720 without the comment? Codes are same. Sorry if i missed something.

Reply:
There is a list of functions named `DeclsToCheckForDeferredDiags`, where callgraph traverse starts in order to emit deferred diagnostics. When the check that I moved was done prior checking that a function is a template, both template and instantiation of function with `sycl_kernel` attribute were getting into `DeclsToCheckForDeferredDiags` list and therefore some of diagnostics that should happen inside function with `sycl_kernel` attribute were emitted twice (see changes in inline-asm.cpp). I've also got some double diagnosing after I made my change in `Sema.cpp`. It seems there was an intention to not add templates to `DeclsToCheckForDeferredDiags` so I've decided ok, let's not do it for SYCL as well, especially if it triggers double diagnosing. I don't remember the reason why we added the code this way though, probably it didn't work for SYCL before big refactoring of deferred diagnostics in the community https://reviews.llvm.org/D70172 . But now we can do like the others and it works.

Reply:
Thanks for the explanation.